azurerm/resource_arm_firewall.go

```diff
@@ -62,24 +62,22 @@ func resourceArmFirewall() *schema.Resource {
 						},
 						"subnet_id": {
 							Type:         schema.TypeString,
-							Required:     true,
+							Optional:     true,
 							ForceNew:     true,
 							ValidateFunc: validateAzureFirewallSubnetName,
 						},
 						"internal_public_ip_address_id": {
-							Type:          schema.TypeString,
-							Optional:      true,
-							Computed:      true,
-							ValidateFunc:  azure.ValidateResourceID,
-							Deprecated:    "This field has been deprecated. Use `public_ip_address_id` instead.",
-							ConflictsWith: []string{"ip_configuration.0.public_ip_address_id"},
+							Type:         schema.TypeString,
+							Optional:     true,
+							Computed:     true,
+							ValidateFunc: azure.ValidateResourceID,
+							Deprecated:   "This field has been deprecated. Use `public_ip_address_id` instead.",
 						},
```

Comment:
Why has the `ConflictsWith` been removed here?

Reply:
yeah, it's a hard decision. 

I don't think we should use `"ip_configuration.0.public_ip_address_id"` here anymore since _ip_configuration_ is a list now.

But I couldn't find a method to deal with the situation, because it needs every element of _ip_configuration_ to check conflicts between the two sub-properties. 

can you offer some suggestions?

Reply:
I would keep the `ConflictsWith` since that will keep the end user mixing and matching the two properties within a single `ip_configuration`. Then I would creating an `ip_configuration` validation function that:
1) Create two bool vars `hasInternal` and `hasPublic`.
2) Loops over all of the `ip_configuration`'s in the schema and look for `internal_public_ip_address_id` and `public_ip_address_id` properties.
a) If it finds an `internal_public_ip_address_id` set the `hasInternal` to `true` and continue the loop.
b) If it finds a `public_ip_address_id` set the `hasPublic` to `true` and continue the loop.
3) Once the code has iterated over all of the `ip_configurations` items in the schema if `hasInternal` is `true` and `hasPublic` is also `true` return an error saying that the `ip_configuration` is invalid and that they must use either one or the other in all defined `ip_configurations` properties.

I would then reference this new function in the `resourceArmFirewallCreateUpdate` function like:

```
func resourceArmFirewallCreateUpdate(d *schema.ResourceData, meta interface{}) error {
	client := meta.(*ArmClient).Network.AzureFirewallsClient
	ctx, cancel := timeouts.ForCreateUpdate(meta.(*ArmClient).StopContext, d)
	defer cancel()

	if err := network.ValidateFirewallConfigurationSettings(d); err != nil {
		return fmt.Errorf(`Error creating Firewall %q (Resource Group %q): the "ip_configuration" is invalid, both "public_ip_address_id" and "internal_public_ip_address_id" have been set, all defined "ip_configuration" blocks must use the same attribute name.`, d.Get("name").(string), d.Get("resource_group_name").(string))
	}
```

Reply:
Thanks for the suggestions. Modified according to your comments.